### PR TITLE
libckteec: Key wrap unwrap

### PR DESCRIPTION
**Pull request by ruchi393**

### CODE REVIEW

File: libckteec/src/pkcs11_api.c [hunk 1]
```diff
@@ -1318,19 +1328,29 @@ CK_RV C_UnwrapKey(CK_SESSION_HANDLE hSession,
...
+		     CKR_DOMAIN_PARAMS_INVALID, CKR_FUNCTION_CANCELED,
+		     CKR_FUNCTION_FAILED, CKR_GENERAL_ERROR, CKR_HOST_MEMORY,
+		     CKR_MECHANISM_INVALID, CKR_MECHANISM_PARAM_INVALID, CKR_OK,
+		     CKR_OPERATION_ACTIVE, CKR_PIN_EXPIRED, CKR_SESSION_CLOSED,
+		     CKR_SESSION_HANDLE_INVALID, CKR_SESSION_READ_ONLY,
+		     CKR_TEMPLATE_INCOMPLETE, CKR_TEMPLATE_INCONSISTENT,
+		     CKR_TOKEN_WRITE_PROTECTED, CKR_USER_NOT_LOGGED_IN,
+		     CKR_UNWRAPPING_KEY_HANDLE_INVALID, CKR_WRAPPED_KEY_INVALID,
+		     CKR_UNWRAPPING_KEY_SIZE_RANGE, CKR_WRAPPED_KEY_LEN_RANGE,
+		     CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT);
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_api.c:1352]: For the 4 last line, prefer keep the alphabetical ordering, as in the spec:
```patch
		     CKR_TEMPLATE_INCOMPLETE, CKR_TEMPLATE_INCONSISTENT,
-		     CKR_TOKEN_WRITE_PROTECTED, CKR_USER_NOT_LOGGED_IN,
-		     CKR_UNWRAPPING_KEY_HANDLE_INVALID, CKR_WRAPPED_KEY_INVALID,
-		     CKR_UNWRAPPING_KEY_SIZE_RANGE, CKR_WRAPPED_KEY_LEN_RANGE,
-		     CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT);
+		     CKR_TOKEN_WRITE_PROTECTED,
+		     CKR_UNWRAPPING_KEY_HANDLE_INVALID,
+		     CKR_UNWRAPPING_KEY_SIZE_RANGE,
+		     CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT,
+		     CKR_USER_NOT_LOGGED_IN, CKR_WRAPPED_KEY_INVALID,
+		     CKR_WRAPPED_KEY_LEN_RANGE);

File: libckteec/src/pkcs11_processing.c [hunk 2]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+	buf += sizeof(wrp_key_handle);
+
+	memcpy(buf, &key_handle, sizeof(key_handle));
+	buf += sizeof(key_handle);
+
+	memcpy(buf, smecha.buffer, smecha.size);
+	buf += smecha.size;
+
+	/* Shm io2: output buffer reference - wrapped key */
+	if (key_len && *key_len) {
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:1450]: no need to test `key_len`

File: libckteec/src/pkcs11_processing.c [hunk 3]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+
+	if (!out_shm) {
+		rv = CKR_HOST_MEMORY;
+		goto bail;
+	}
+
+	rv = ckteec_invoke_ctrl_out(PKCS11_CMD_WRAP_KEY, ctrl, out_shm,
+				    &out_size);
+
+	if (key_len && (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL))
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:1466]: no need to test `key_len`

File: libckteec/src/pkcs11_processing.c [hunk 4]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+	struct serializer sattr = { 0 };
+	uint32_t session_handle = session;
+	uint32_t obj_handle = unwrapping_key;
+	size_t ctrl_size = 0;
+	uint32_t key_handle = 0;
+	char *buf = NULL;
+	size_t out_size = 0;
+
+	if (!handle || !mechanism || (count && !attribs) ||
+	    (!unwrapping_key) || (wrapped_key_len && !wrapped_key))
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:1498]: can remove parentheses around `unwrapping_key`

File: libckteec/src/pkcs11_processing.c [hunk 5]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+	if (key_len && (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL))
+		*key_len = out_size;
+
+	if (rv == CKR_BUFFER_TOO_SMALL && out_size && !wrapped_key)
+		rv = CKR_OK;
+
+bail:
+	ckteec_free_shm(out_shm);
+	ckteec_free_shm(ctrl);
+
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:1475]: release_serial_object(&smecha);

**Review by etienne-lms — COMMENTED** [review]
Review comments for "libckteec: Add implementation for wrapping/unwrapping keys".

`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "libckteec: Add missing indirect template"
* "libckteec: Add missing error codes"
* "libckteec: Allocate command ID's for wrapping/unwrapping keys"
  (s/ID's/IDs/ in commit log)
* "libckteec: Correct error code in C_Sign()"
  Maybe add tag  `Fixes: 3f44b8702995 ("libckteec: add support for Signing and Verification Functions") `

**Comment by ruchi393** [comment]
Thanks @etienne-lms . I have pushed changes to take care of review comments and error reported by Travis.

File: libckteec/src/pkcs11_processing.c (hunk 5 above)
Comment (ruchi393) [reply libckteec/src/pkcs11_processing.c:1475]: My bad. I have added it now.

File: libckteec/src/pkcs11_processing.c (hunk 4 above)
Comment (ruchi393) [reply libckteec/src/pkcs11_processing.c]: Done. Travis also complained about it .

File: libckteec/src/pkcs11_processing.c (hunk 3 above)
Comment (ruchi393) [reply libckteec/src/pkcs11_processing.c]: You are right. We have already checked it at beginning of the function. Have removed the check.

File: libckteec/src/pkcs11_processing.c (hunk 2 above)
Comment (ruchi393) [reply libckteec/src/pkcs11_processing.c]: Removed

File: libckteec/src/pkcs11_api.c (hunk 1 above)
Comment (ruchi393) [reply libckteec/src/pkcs11_api.c]: Changed. I was trying to optimize space usage :-). But you are right better to keep the same order. Makes comparison with spec easier.

File: libckteec/src/pkcs11_processing.c [hunk 6]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+CK_RV ck_wrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		  CK_OBJECT_HANDLE wrapping_key, CK_OBJECT_HANDLE key,
+		  CK_BYTE_PTR wrapped_key, CK_ULONG_PTR key_len)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	TEEC_SharedMemory *ctrl = NULL;
+	TEEC_SharedMemory *out_shm = NULL;
+	struct serializer smecha = { 0 };
+	uint32_t session_handle = session;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:1405]: CK_SESSION_HANDLE is CK_ULONG and that is 64 bits in 64 bit arch and does not fit in 32 uint32_t without a cast.

File: libckteec/src/pkcs11_processing.c [hunk 7]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+	uint32_t wrp_key_handle = wrapping_key;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:1406]: `CK_ULONG` issue here too

File: libckteec/src/pkcs11_processing.c [hunk 8]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+	uint32_t key_handle = key;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:1407]: And here too.

File: libckteec/src/pkcs11_processing.c [hunk 9]
```diff
@@ -1393,3 +1393,179 @@ CK_RV ck_release_active_processing(CK_SESSION_HANDLE session,
...
+	buf += sizeof(session_handle);
+
+	memcpy(buf, &wrp_key_handle, sizeof(wrp_key_handle));
+	buf += sizeof(wrp_key_handle);
+
+	memcpy(buf, &key_handle, sizeof(key_handle));
+	buf += sizeof(key_handle);
+
+	memcpy(buf, smecha.buffer, smecha.size);
+	buf += smecha.size;
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:1447]: I understand where you come from ;) -- but this is not needed.

File: libckteec/src/pkcs11_processing.c (hunk 6 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:1405]: We do have `serialize_ck_ulong()` but I suppose that is a bit off but we could have helper for `CK_ULONG`.

**Comment by vesajaaskelainen** [comment]
Btw -- do we have CI build for OS side with 64 bit compiler? Because in theory it should have noticed that 64 bits do not fit in 32 bits.

**Comment by vesajaaskelainen** [comment]
Btw. please tune up the PR's title -- we do not have TA's in optee client.

File: libckteec/src/pkcs11_processing.c (hunk 6 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:1405]: I suppose it is OK to go with the existing way before the master is fixed for all of those?

**Comment by ruchi393** [comment]
> Btw. please tune up the PR's title -- we do not have TA's in optee client.

Changed. Thanks for noticing this.

File: libckteec/src/pkcs11_processing.c (hunk 9 above)
Comment (ruchi393) [reply libckteec/src/pkcs11_processing.c:1447]: Oh yes. There is nothing more to add in the buffer, so shifting this pointer is not required. I will remove this.

File: libckteec/include/pkcs11_ta.h [hunk 10]
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
...
+	 *              32bit wrapping key handle,
+	 *              32bit key handle,
+	 *              (struct pkcs11_attribute_head)mechanism + mecha params
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [out] memref[2] = wrapped key
+	 *
+	 * This command relates to the PKCS#11 API function C_WrapKey().
+	 */
+	PKCS11_CMD_WRAP_KEY = 50,
```
Comment (vesajaaskelainen) [inline libckteec/include/pkcs11_ta.h:675]: +1

File: libckteec/include/pkcs11_ta.h [hunk 11]
```diff
@@ -657,6 +657,40 @@ enum pkcs11_ta_cmd {
...
+	 *              (struct pkcs11_attribute_head)mechanism + mecha params,
+	 *              (struct pkcs11_object_head)attribs + attributes data
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [in]  memref[1] = wrapped key
+	 * [out] memref[2] = 32bit object handle
+	 *
+	 * This command relates to the PKCS#11 API function C_UnwrapKey().
+	 */
+	PKCS11_CMD_UNWRAP_KEY = 51,
```
Comment (vesajaaskelainen) [inline libckteec/include/pkcs11_ta.h:693]: +1

**Comment by ruchi393** [comment]
Rebased, squashed commits, tested again and updated the PR.

@etienne-lms , @vesajaaskelainen , please help review.

File: libckteec/src/pkcs11_processing.c [hunk 12]
```diff
@@ -1520,3 +1520,178 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
...
+		rv = CKR_HOST_MEMORY;
+		goto bail;
+	}
+
+	buf = ctrl->buffer;
+
+	memcpy(buf, &session_handle, sizeof(session_handle));
+	buf += sizeof(session_handle);
+
+	memcpy(buf, &obj_handle, sizeof(obj_handle));
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:1654]: `obj_handle` as a name do not have any relation that it is unwrapping key.

File: libckteec/src/pkcs11_processing.h [hunk 13]
```diff
@@ -127,4 +127,13 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
...
 			   CK_OBJECT_HANDLE_PTR priv_key);
 
+CK_RV ck_wrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		  CK_OBJECT_HANDLE wrapping_key, CK_OBJECT_HANDLE key,
+		  CK_BYTE_PTR wrapped_key, CK_ULONG_PTR key_len);
+
+CK_RV ck_unwrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		    CK_OBJECT_HANDLE unnwrapping_key, CK_BYTE_PTR wrapped_key,
+		    CK_ULONG key_len, CK_ATTRIBUTE_PTR attribs, CK_ULONG count,
+		    CK_OBJECT_HANDLE_PTR handle);
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.h:137]: In implementation you have nicer variable names -- please use them here too:
```
CK_RV ck_unwrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
		    CK_OBJECT_HANDLE unwrapping_key, CK_BYTE_PTR wrapped_key,
		    CK_ULONG wrapped_key_len, CK_ATTRIBUTE_PTR attribs,
		    CK_ULONG count, CK_OBJECT_HANDLE_PTR handle)
```

File: libckteec/src/pkcs11_processing.c [hunk 14]
```diff
@@ -1520,3 +1520,178 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
 
 	return rv;
 }
+
+CK_RV ck_wrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		  CK_OBJECT_HANDLE wrapping_key, CK_OBJECT_HANDLE key,
+		  CK_BYTE_PTR wrapped_key, CK_ULONG_PTR key_len)
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:1526]: /key_len/wrapped_key_len/

File: libckteec/src/pkcs11_processing.h [hunk 15]
```diff
@@ -127,4 +127,13 @@ CK_RV ck_generate_key_pair(CK_SESSION_HANDLE session,
 			   CK_OBJECT_HANDLE_PTR pub_key,
 			   CK_OBJECT_HANDLE_PTR priv_key);
 
+CK_RV ck_wrap_key(CK_SESSION_HANDLE session, CK_MECHANISM_PTR mechanism,
+		  CK_OBJECT_HANDLE wrapping_key, CK_OBJECT_HANDLE key,
+		  CK_BYTE_PTR wrapped_key, CK_ULONG_PTR key_len);
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.h:132]: /key_len/wrapped_key_len/

**Comment by vesajaaskelainen** [comment]
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`
- https://github.com/OP-TEE/optee_client/pull/265/commits/77408ed6798651b8f7d0c9d568575a81f188ebc5 - "libckteec: Add missing indirect template"
- https://github.com/OP-TEE/optee_client/pull/265/commits/f25a836f1609cdce9da5698067914a5893e39e9b - "libckteec: Add missing error codes"
- https://github.com/OP-TEE/optee_client/pull/265/commits/70a6b85175ec4b02340a4a3a5de8c7f5413acdfb - "libckteec: Allocate command ID's for wrapping/unwrapping keys"
- https://github.com/OP-TEE/optee_client/pull/265/commits/ec152f565049d25f19629322266619b1049f99aa - "libckteec: Correct error code in C_Sign()"

And with those minor things fixed:
- https://github.com/OP-TEE/optee_client/pull/265/commits/c95b48e5f0dc9004cca46b5cd95eaa0011633806 - "libckteec: Add implementation for wrapping/unwrapping keys"

**Comment by ruchi393** [comment]
Added tags as suggested.

Fixed "libckteec: Add implementation for wrapping/unwrapping keys" based on comments above.

Thanks @vesajaaskelainen for the reviewing.

**Comment by vesajaaskelainen** [comment]
For commit:
- https://github.com/OP-TEE/optee_client/pull/265/commits/517ed320728114e68eb1d4aee2a71d536a8367eb - libckteec: Add implementation for wrapping/unwrapping keys

`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

**Comment by ruchi393** [comment]
Rebased and tags applied.

**Comment by jforissier** [comment]
I think this is safe to merge even though the `optee_os` part is still being reviewed, isn't it?

**Comment by ruchi393** [comment]
> I think this is safe to merge even though the `optee_os` part is still being reviewed, isn't it?

I would agree,we can merge this now.

**Pull request merged**